File: tools/bench/xi4_work_stealing.py
"""워크 스틸링에서 일이 실제로 어떻게 나뉘는가 — XI-4.

작업 2만 건을 워커 0의 덱에만 몰아 두고 워커 4개(코어 4)를 동시에 출발시킨다.
훔치기가 동작하면 넷이 나눠 갖는다. GIL 이 있는 Python 에서는 나눠 가지지 못하는
경우가 흔하다는 것을 워커별 처리 건수로 본다.

측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / CPython 3.13.12 / 4코어)

사용법:
    python3.13 tools/bench/xi4_work_stealing.py
"""

import statistics
import sys
import threading
from collections import deque

N_WORKERS, N_TASKS = 4, 20000
RUNS = 10


def one_run():
    decks = [deque() for _ in range(N_WORKERS)]
    locks = [threading.Lock() for _ in range(N_WORKERS)]
    decks[0].extend(range(N_TASKS))
    go = threading.Event()
    processed = [0] * N_WORKERS

    def take_own(w):
        with locks[w]:
            return decks[w].pop() if decks[w] else None

    def steal(w):
        for k in range(1, N_WORKERS):
            v = (w + k) % N_WORKERS
            with locks[v]:
                if decks[v]:
                    return decks[v].popleft()
        return None

    def worker(w):
        go.wait()
        while True:
            job = take_own(w)
            if job is None:
                job = steal(w)
                if job is None:
                    return
            processed[w] += 1

    ws = [threading.Thread(target=worker, args=(w,)) for w in range(N_WORKERS)]
    for w in ws:
        w.start()
    go.set()
    for w in ws:
        w.join()
    return processed


def main():
    print(f"python {sys.version.split()[0]} · GIL={sys._is_gil_enabled()} · "
          f"워커 {N_WORKERS} · 코어 4 · 작업 {N_TASKS:,}건 · {RUNS}회 실행")
    shares = []
    for r in range(RUNS):
        p = one_run()
        assert sum(p) == N_TASKS
        share = max(p) / N_TASKS * 100
        shares.append(share)
        print(f"  {r + 1:2}회차 워커별 {p} · 최다 워커 몫 {share:.0f}%")
    print(f"최다 워커 몫 중앙값 {statistics.median(shares):.0f}% "
          f"(범위 {min(shares):.0f} ~ {max(shares):.0f}%) · 이상적 분배는 25%")


if __name__ == "__main__":
    main()

File: tools/bench/xi4_work_stealing.cpp
// 워크 스틸링에서 일이 실제로 어떻게 나뉘는가 (C++) — XI-4.
//
// tools/bench/xi4_work_stealing.py 와 같은 실험. 작업 2만 건을 워커 0의 덱에만
// 몰아 두고 워커 4개(코어 4)를 동시에 출발시킨다.
//
// 측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / g++ 13.3.0 -std=c++17 -O2 / 4코어)
// 빌드: g++ -std=c++17 -O2 -pthread tools/bench/xi4_work_stealing.cpp -o /tmp/xi4w && /tmp/xi4w

#include <algorithm>
#include <atomic>
#include <cstdio>
#include <deque>
#include <mutex>
#include <thread>
#include <vector>
using namespace std;

const int N_WORKERS = 4, N_TASKS = 20000;
const int RUNS = 10;

struct Deck {
    deque<int> d;
    mutex mtx;
};

vector<int> one_run() {
    vector<Deck> decks(N_WORKERS);
    for (int i = 0; i < N_TASKS; i++) decks[0].d.push_back(i);
    atomic<bool> go{false};
    vector<int> processed(N_WORKERS, 0);

    vector<thread> ws;
    for (int w = 0; w < N_WORKERS; w++)
        ws.emplace_back([&, w] {
            while (!go.load()) this_thread::yield();
            while (true) {
                int job = -1;
                {
                    lock_guard<mutex> g(decks[w].mtx);
                    if (!decks[w].d.empty()) {
                        job = decks[w].d.back();
                        decks[w].d.pop_back();
                    }
                }
                if (job < 0) {
                    for (int k = 1; k < N_WORKERS && job < 0; k++) {
                        Deck& v = decks[(w + k) % N_WORKERS];
                        lock_guard<mutex> g(v.mtx);
                        if (!v.d.empty()) {
                            job = v.d.front();
                            v.d.pop_front();
                        }
                    }
                }
                if (job < 0) return;
                processed[w]++;
            }
        });
    go.store(true);
    for (auto& t : ws) t.join();
    return processed;
}

int main() {
    printf("워커 %d · 코어 4 · 작업 %d건 · %d회 실행\n", N_WORKERS, N_TASKS, RUNS);
    vector<double> shares;
    for (int r = 0; r < RUNS; r++) {
        vector<int> p = one_run();
        int sum = 0, mx = 0;
        for (int v : p) { sum += v; mx = max(mx, v); }
        double share = 100.0 * mx / N_TASKS;
        shares.push_back(share);
        printf("  %2d회차 워커별 [%d, %d, %d, %d] · 합 %d · 최다 워커 몫 %.0f%%\n",
               r + 1, p[0], p[1], p[2], p[3], sum, share);
    }
    sort(shares.begin(), shares.end());
    printf("최다 워커 몫 중앙값 %.0f%% (범위 %.0f ~ %.0f%%) · 이상적 분배는 25%%\n",
           shares[shares.size() / 2], shares.front(), shares.back());
    return 0;
}
